### ISSUE TITLE
Bug with useTransition and the leave method being run more than once

### ISSUE THREAD

**Issue by parkroolucas** [react-spring/react-spring#593]
Please refer to my CodeSandbox demonstration of the issue here: [https://codesandbox.io/s/y8w191jvx](https://codesandbox.io/s/y8w191jvx)

Clicking on the buttons a few times will show you what I mean. Elements that are meant to have toggled out come jumping back to full height, and then disappear once again.

I have found that the reason for this is that the `leave` method is being re-run on all unmounting elements when `selectedIndexes` is updated whilst they are in the process of leaving. Because the first part of my `leave` method has a `duration` of 0, it causes the snap effect back to full height (my justification for doing this is to properly animate away from a height of `auto`, as you can see).

Is this a bug with `react-spring`, or simply my fault for misusing / abusing the API?

Thanks in advance.

**Comment by drcmda** [comment]
no, it's a feature. 😄for real, though. it makes sure an item can go in and out, even with the same key, and be transitioned out safely. it does that by calculating arbitrary keys, that's why you need to use these in the view instead of your own.

if you don't want that you can do: unique: true.

now if item A enters, you remove it, and while it's fading out you enter A again, it will trash the ongoing out-transition and move the item back to enter state. If you need it to start from scratch you can combine it with the reset: true flag, if you A starts with opacity: 0, but has faded out to 0.5 already, when it comes in it will start at 0 again.

**Comment by parkroolucas** [comment]
What I'm trying to do is create a toggler, where the expanding element can be toggled in and out. My reason for setting `unique: true` is so that the expanding element isn't created multiple times if toggled many times in succession.

Additionally, what I'm trying to do is, upon entry:

1. Start the expanding element at height 0.
2. Find the content height of the expanding element, and animate it to that height.
3. Once the animation is done, change the height to `auto`, so that the inner contents of the expanding element can be changed, and the expanding element can readjust to suit that height. (I want to do this instantly, because it's not an actual height change

Upon exit (leave):

1. Find the content height of the expanding height, and change it from `auto` to that particular height, so that `react-spring` can animate the height from that height to 0. (I want to do this instantly, because it's not an actual height change, but just to enable animation in `react-spring`)
2. Animate the height to 0.

But as you can see, I'm getting problems when clicking on multiple togglers in quick succession, because the first state of the leave (change the height from `auto` to that particular height) is cancelling out the fact that it was already well on its way to animating towards 0; the end result is that it goes from the original height, to halfway towards 0, and then snaps back to its original height, and then back to 0 again, and so on if more buttons are clicked.

I hope that makes sense :) Could you provide any suggestion on how I could fix this? Thank you for your time.

**Comment by drcmda** [comment]
maybe its better to interpolate auto in the view instead of making a chain. chains are async and harder to control. 

div style={{ height: props.height.interpolate(h => h === condition ? 'auto' : h) }}

This example uses something like it, though theres q resize observer in there.

**Comment by parkroolucas** [comment]
Thank you for the wise words - that's definitely helpful.

However, I don't see how that solves the second part of my problem here:

1. Find the content height of the expanding height, and change it from `auto` to that particular height, so that `react-spring` can animate the height from that height to 0. (I want to do this instantly, because it's not an actual height change, but just to enable animation in `react-spring`)
2. Animate the height to 0.

It seems to me like interpolating `auto` will do the following:

1. Use pixel height until the element reaches the full `offsetHeight` value.
2. Once the element reaches the full `offsetHeight` value, replace pixel height with `auto`.
3. However, if the height of the content changes during this time and before leave, then we get a jittery animation on leave (say the element was originally 400px tall, but then its content makes it 500px tall... wouldn't the animation start from 400px nonetheless?)

How would I address this issue?

**Comment by jcousins-ynap** [comment]
@aleclarson why was this closed and labelled "invalid"?

**Comment by aleclarson** [comment]
According to https://github.com/react-spring/react-spring/issues/593#issuecomment-472827872, the original issue is by design, hence the invalid label. If you or @lucaschen have another question or suggestion, please open a separate issue. Thanks!
